src/core/meshcop/joiner.cpp

```diff
@@ -197,19 +201,31 @@ void Joiner::HandleDiscoverResult(otActiveScanResult *aResult)
         steeringData.SetLength(aResult->mSteeringData.mLength);
         memcpy(steeringData.GetValue(), aResult->mSteeringData.m8, steeringData.GetLength());
 
-        if (steeringData.DoesAllowAny() ||
-            (steeringData.GetBit(mCcitt % steeringData.GetNumBits()) &&
-             steeringData.GetBit(mAnsi % steeringData.GetNumBits())))
+        if (steeringData.DoesAllowAny())
         {
-            mJoinerUdpPort = aResult->mJoinerUdpPort;
-            mJoinerRouterPanId = aResult->mPanId;
-            mJoinerRouterChannel = aResult->mChannel;
-            memcpy(&mJoinerRouter, &aResult->mExtAddress, sizeof(mJoinerRouter));
+            VerifyOrExit(!mJoinerRouterIsSpecific);
+            VerifyOrExit(aResult->mLqi > mJoinerRouterLqi);
+        }
+        else if (steeringData.GetBit(mCcitt % steeringData.GetNumBits()) &&
+                 steeringData.GetBit(mAnsi % steeringData.GetNumBits()))
+        {
+            if (mJoinerRouterIsSpecific)
+            {
+                VerifyOrExit(aResult->mLqi > mJoinerRouterLqi);
+            }
         }
         else
         {
             otLogDebgMeshCoP(GetInstance(), "Steering data does not include this device");
+            ExitNow();
         }
+
+        mJoinerRouterIsSpecific = !steeringData.DoesAllowAny();
```

Comment:
Minor optimization, could add `mJoinerRouterIsSpecific = true` after L215 above and avoid another call to `DoesAllowAny()`.

Reply:
Done